### PR TITLE
feat(connlib): respond w/ ICMP unreachable on socket error

### LINKED ISSUE #6353: Don't use IPv6 DNS upstreams when there's no IPv6 connectivity
If an admin specifies an IPv6 upstream, clients without IPv6 will fail to connect to it and time DNS queries out. The OS (macOS in this case) may never mark that server as unresponsive, causing DNS resolution to fail for the user.

Source: https://firezonehq.slack.com/archives/C069H865MHP/p1724093936014149

### PR DESCRIPTION
**Pull request by thomaseizinger**
Whenever `connlib` receives an IP packet from the TUN interface, we pass it to our state machine for processing. In most cases, the outcome of this processing is a `Transmit`: The encrypted WG packet that needs to be sent to the gateway.

In case the IP packet was a DNS query, we might forward it to one of our DNS servers. Sending such a packet can fail with "Destination Unreachable". This is an ICMP error code that is forwarded to us by the operating system. To let the application know that this DNS server is not routable, we now respond with the same ICMP error code back to the application via the TUN interface.

Resolves: #6353. (Hopefully)

### CODE REVIEW

File: rust/connlib/tunnel/src/lib.rs [hunk 1]
```diff
@@ -139,11 +139,23 @@ impl ClientTunnel {
...
+                    match self.io.send_network(transmit) {
+                        Ok(()) => {}
+                        Err(e) if is_network_unreachable(&e) => {
+                            let response =
+                                ip_packet::make::icmp_network_unreachable(dst, src).unwrap();
+
+                            self.io.send_device(response)?;
+                        }
+                        Err(e) => return Poll::Ready(Err(e)),
+                    }
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/lib.rs:158]: This is the same codepath that we also use for sending encrypted WG packets. We should probably add a safe-guard here to ensure we only do this for forwarded DNS queries in order to not disrupt any data plane connections during roaming.

Comment (thomaseizinger) [reply rust/connlib/tunnel/src/lib.rs:158]: I think the way I want to solve this is by passing the `io::Error` to the state instead, together with the destination of the datagram.

We can then internally look up whether or not that was a forwarded DNS query and if it is, we can construct an ICMP error. That should be more robust and most importantly, can be tested as part of tunnel test.

**Review by jamilbk — COMMENTED** [review]
Yeah I don't think it can hurt to pass ICMP messages along properly. No comment on the approach, will defer to others.

**Comment by thomaseizinger** [comment]
> Yeah I don't think it can hurt to pass ICMP messages along properly. No comment on the approach, will defer to others.

I am hoping for this to resolve #6353 without us having to do any filtering of DNS servers.

**Comment by thomaseizinger** [comment]
I really want to test this using `tunnel_test` but our current `ip-packet` crate is a bit of a mess. I've started to clean that up by replacing it bit-by-bit with `etherparse` in #6502.

**Comment by thomaseizinger** [comment]
@conectado Do you have some commits that could be pushed onto this PR?

**Comment by conectado** [comment]
> @conectado Do you have some commits that could be pushed onto this PR?

I quickly threw #6654 together, to see if it fixed the bug but since we never managed to reproduce I didn't clean it up.

Do we think it's worth adding this even if we can't reproduce the original bug?

**Comment by thomaseizinger** [comment]
> > @conectado Do you have some commits that could be pushed onto this PR?
> 
> I quickly threw #6654 together, to see if it fixed the bug but since we never managed to reproduce I didn't clean it up.
> 
> Do we think it's worth adding this even if we can't reproduce the original bug?

I guess we can always come back to this code if we ever need it. I am still a bit uneasy about not responding to those DNS queries at all.

**Pull request closed without merging**